src/ray/raylet/worker_pool.cc

```diff
@@ -587,68 +577,82 @@ void WorkerPool::PushWorker(const std::shared_ptr<WorkerInterface> &worker) {
+      continue;
+    }
 
-  auto worker_id = worker->WorkerId();
-  const auto pid = worker->GetProcess().GetId();
-  if (worker_state.idle.count(worker) == 0) {
-    return;
-  }
-  if (worker_state.starting_worker_processes.count(worker->GetProcess()) > 0) {
-    // A Java worker process may hold multiple workers.
-    RAY_LOG(DEBUG) << "Some workers of pid " << pid
-                   << " are pending registration. Skip killing worker " << worker_id;
-    return;
-  }
+    auto &worker_state = GetStateForLanguage((*it)->GetLanguage());
 
-  // Make sure all workers in this worker process are idle.
-  // This block of code is needed by Java workers.
-  std::unordered_set<std::shared_ptr<WorkerInterface>> workers_in_the_same_process;
-  for (const auto &worker_in_the_same_process : worker_state.registered_workers) {
-    if (worker_in_the_same_process->GetProcess().GetId() == pid) {
-      if (worker_state.idle.count(worker_in_the_same_process) == 0) {
-        // Another worker in this process isn't idle, so this process can't be killed.
-        return;
-      } else {
+    if (worker_state.starting_worker_processes.count(process) > 0) {
+      // A Java worker process may hold multiple workers.
+      // Some workers of this process are pending registration. Skip killing this worker.
+      continue;
+    }
+
+    // Make sure all workers in this worker process are idle.
+    // This block of code is needed by Java workers.
+    std::unordered_set<std::shared_ptr<WorkerInterface>> workers_in_the_same_process;
+    bool can_be_killed = true;
+    for (const auto &worker_in_the_same_process : worker_state.registered_workers) {
+      if (worker_in_the_same_process->GetProcess().GetId() == process.GetId()) {
+        if (worker_state.idle.count(worker_in_the_same_process) == 0) {
+          // Another worker in this process isn't idle, so this process can't be killed.
+          can_be_killed = false;
+          break;
+        }
         workers_in_the_same_process.insert(worker_in_the_same_process);
       }
     }
+    if (!can_be_killed) {
+      continue;
+    }
+
+    for (auto worker_it = workers_in_the_same_process.begin();
+         worker_it != workers_in_the_same_process.end(); worker_it++) {
+      RAY_LOG(INFO) << "The worker pool has " << running_size
+                    << " registered workers which exceeds the soft limit of "
+                    << num_workers_soft_limit_ << ", and worker "
+                    << (*worker_it)->WorkerId() << " with pid " << process.GetId()
+                    << " is idle. Kill it.";
+      // Remove the worker from the idle pool so it can't be popped anymore. However, we
+      // don't remove it from the registered pool because we want the worker to go through
+      // the normal disconnection logic in Node Manager.
+      RemoveWorker(worker_state.idle, *worker_it);
+      RAY_CHECK(pending_unregistration_workers.insert(*worker_it).second);
+      running_size--;
+    }
```

Comment:
Same comment here, can this be a utility method of the process class?

Reply:
I'm not sure what should be moved to the process class. Can you be more specific?

Reply:
Instead of for loops, it would be much more readable to have `process->KillAllWorkers()` or `process->HasIdleWorkers()` utility methods to reduce the size of the TryKillingIdleWorkers function.

Reply:
It sounds weird to me. `Process` is a low-level class that shouldn't be aware of the worker concept. and it's used in other places as well. BTW, we will change the implementation from killing workers to notify them to exit gracefully in the enabling PR: https://github.com/ray-project/ray/pull/10570/files#diff-64c50c0cc29e8f9f96ebf5df2b668f1c